jointMAF <- function(mafdf, CSQ_info, vcfMain){
    mafFinal <- cbind(mafdf, Allele = CSQ_info$Allele, Gene = CSQ_info$Gene, 
        Feature = CSQ_info$Feature, Feature_type = CSQ_info$Feature_type,
        One_Consequence=CSQ_info$Consequence,Consequence = CSQ_info$Consequence,
        cDNA_position=CSQ_info$cDNA_position,CDS_position=CSQ_info$CDS_position,
        Protein_position = CSQ_info$Protein_position,
        Amino_acids = CSQ_info$Amino_acids, Codons = CSQ_info$Codons,
        Existing_variation = getValue(CSQ_info$Existing_variation),
        ALLELE_NUM = getValue(CSQ_info$ALLELE_NUM),DISTANCE = CSQ_info$DISTANCE, 
        TRANSCRIPT_STRAND = CSQ_info$STRAND, SYMBOL = CSQ_info$SYMBOL, 
        SYMBOL_SOURCE = CSQ_info$SYMBOL_SOURCE, HGNC_ID = CSQ_info$HGNC_ID, 
        BIOTYPE = CSQ_info$BIOTYPE, CANONICAL = getValue(CSQ_info$CANONICAL),
        CCDS = getValue(CSQ_info$CCDS), ENSP = getValue(CSQ_info$ENSP), 
        SWISSPROT=getValue(CSQ_info$SWISSPROT),TREMBL=getValue(CSQ_info$TREMBL), 
        UNIPARC = getValue(CSQ_info$UNIPARC),RefSeq = getValue(CSQ_info$RefSeq), 
        SIFT = getValue(CSQ_info$SIFT), PolyPhen = getValue(CSQ_info$PolyPhen), 
        Exon = CSQ_info$EXON, INTRON = CSQ_info$INTRON, 
        DOMAINS = getValue(CSQ_info$DOMAINS), GMAF=getValue(CSQ_info$AF), 
        AFR_MAF = getValue(CSQ_info$AFR_AF),AMR_MAF = getValue(CSQ_info$AMR_AF), 
        ASN_MAF=getValue(CSQ_info$ASN_MAF), EAS_MAF = getValue(CSQ_info$EAS_AF),
        EUR_MAF = getValue(CSQ_info$EUR_AF),SAS_MAF = getValue(CSQ_info$SAS_AF), 
        AA_MAF = getValue(CSQ_info$AA_AF), EA_MAF = getValue(CSQ_info$EA_AF), 
        CLIN_SIG=getValue(CSQ_info$CLIN_SIG),SOMATIC=getValue(CSQ_info$SOMATIC), 
        PUBMED = CSQ_info$PUBMED,MOTIF_NAME=getValue(CSQ_info$MOTIF_NAME),
        MOTIF_POS = getValue(CSQ_info$MOTIF_POS),
        HIGH_INF_POS = getValue(CSQ_info$HIGH_INF_POS),
        MOTIF_SCORE_CHANGE = getValue(CSQ_info$MOTIF_SCORE_CHANGE),
        IMPACT = CSQ_info$IMPACT, PICK = getValue(CSQ_info$PICK),
        VARIANT_CLAS=getValue(CSQ_info$VARIANT_CLASS),
        TSL=getValue(CSQ_info$TSL),HGVS_OFFSET = getValue(CSQ_info$HGVS_OFFSET),
        PHENO = getValue(CSQ_info$PHENO), MINIMISED = '', 
        gnomAD_AF = getValue(CSQ_info$gnomAD_AF),
        gnomAD_AFR_AF = getValue(CSQ_info$gnomAD_AFR_AF),
        gnomAD_AMR_AF = getValue(CSQ_info$gnomAD_AMR_AF),
        gnomAD_ASJ_AF = getValue(CSQ_info$gnomAD_ASJ_AF),
        gnomAD_EAS_AF = getValue(CSQ_info$gnomAD_EAS_AF),
        gnomAD_FIN_AF = getValue(CSQ_info$gnomAD_FIN_AF),
        gnomAD_NFE_AF = getValue(CSQ_info$gnomAD_NFE_AF),
        gnomAD_OTH_AF = getValue(CSQ_info$gnomAD_OTH_AF),
        gnomAD_SAS_AF = getValue(CSQ_info$gnomAD_SAS_AF),
        ExAC_AF = getValue(CSQ_info$ExAC_AF),
        ExAC_AF_adj = getValue(CSQ_info$ExAC_AF_adj),
        ExAC_AF_AFR = getValue(CSQ_info$ExAC_AF_AFR),
        ExAC_AF_AMR = getValue(CSQ_info$ExAC_AF_AMR),
        ExAC_AF_EAS = getValue(CSQ_info$ExAC_AF_EAS),
        ExAC_AF_FIN = getValue(CSQ_info$ExAC_AF_FIN),
        ExAC_AF_NFE = getValue(CSQ_info$ExAC_AF_NFE),
        ExAC_AF_OTH = getValue(CSQ_info$ExAC_AF_OTH),
        ExAC_AF_SAS = getValue(CSQ_info$ExAC_AF_SAS),
        GENE_PHENO = getValue(CSQ_info$GENE_PHENO), FILTER = vcfMain$FILTER)}

getValue <- function(chars){
    if (is.null(chars)){
      return("")
    }else{
      return(chars)
    }
}
