#' mutFilterType
#' @description Filter variants of certain types
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' @param keepType A group of variant classifications will be kept,
#' including 'exonic', 'nonsynonymous' and 'ALL'. Default: 'exonic'.
#'
#' @return An MAF data frame after variant type filtration.
#'
#' @export mutFilterType
#' @examples
#' maf <- vcfToMAF(system.file("extdata", "GC48-2_mutect2.vep.vcf",
#' package = "CaMutQC"))
#' mafF <- mutFilterType(maf)


mutFilterType <- function(maf, keepType = 'exonic') {

  # process keepType
  if (keepType == 'exonic') {
    flterType <- c('RNA', 'Intron', 'IGR', '5\'Flank', '3\'Flank',
               '5\'UTR', '3\'UTR')
  } else if (keepType == 'nonsynonymous') {
    flterType <- c("3'UTR", "5'UTR", "3'Flank", "Targeted_Region", "Silent",
               "Intron", "RNA", "IGR", "Splice_Region", "5'Flank", "lincRNA",
               "De_novo_Start_InFrame", "De_novo_Start_OutOfFrame",
               "Start_Codon_Ins", "Start_Codon_SNP", "Stop_Codon_Del")
  } else if (keepType == 'ALL'){
    return(maf)
  } else {
    stop('Please select one category of types from exonic, nonsynonymous and ALL.')
  }

  # add T tag
  tags <- rownames(maf[maf$Variant_Classification %in% flterType, ])
  maf[tags, 'CaTag'] <- paste0(maf[tags, 'CaTag'], 'T')

  return(maf)
}
