#' mutFilterReg
#' @description Filter variants in specific regions.
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' @param bedFile A bed file that contains region information.
#' Default: NULL
#' @param bedFilter Whether to filter the information in bed file or not, which
#' only leaves segments in Chr1-Ch22, ChrX and ChrY. Default: TRUE
#'
#' @return An MAF data frame where some variants
#' have R tag in CaTag column for region filtration.
#'
#' @export mutFilterReg
#' @examples
#' maf <- vcfToMAF(system.file("extdata", "GC48-2_mutect2.vep.vcf",
#' package = "CaMutQC"))
#' mafF <- mutFilterReg(maf, bedFile = system.file("extdata/bed",
#' "Twist_Exome_Target_hg19.bed",package = "CaMutQC"))


mutFilterReg <- function(maf, bedFile = NULL, bedFilter = TRUE){

  if (is.null(bedFile)){
    message('No bed file detected, so no variants will get an R flag.')
  }else{
    # bed <- read.table('../Data/Twist_Exome_Target_hg19.bed')
    bed <- read.table(bedFile)
    chromVaild <- c(paste0('chr', 1:22), 'chrX', 'chrY')

    ## check for validation of bed file
    if (nrow(bed) < 3 | any(substring(bed$V1, 1,3) != 'chr') |
        typeof(bed$V2) != 'integer' | typeof(bed$V3) != 'integer'){
      stop(paste0('Invaild bed file. Please input vaild bed file',
                  ' with at least 3 columns of porper type.'))  }

    if(any(!(bed$V1 %in% chromVaild))){
      if (!bedFilter){
        stop(paste0('Invaild chrom in bed File. Please turn on bedFiler',
                    ' if you want to filter the bed file'))
      } else {
        bed <- bed[which(bed$V1 %in% chromVaild), ]
      }
    }

    ## sort bed object
    bedProc <- unique(bed[, 1:3])
    colnames(bedProc) <- c('chrom', 'chromStart', 'chromEnd')

    ## process maf data and start targeting
    mafTar <- maf[, c("Chromosome", "Start_Position", "End_Position")]
    chrs <- unique(maf$Chromosome)
    # nums <- rep(0, length(chrs))
    allTar <- list()
    for (c in seq_len(length(chrs))) {
      mafSepc <- mafTar[which(mafTar$Chromosome == chrs[c]), ]
      bedSepc <- bedProc[which(bedProc$chrom == chrs[c]), ]
      l <- list()
      for(i in seq_len(nrow(bedSepc))) {
        l[[i]] <- mutRegionTag(mafSepc, bedSepc[i, ])
      }
      allTar[[c]] <- l
    }

    # get complement subset (variants not in bed region)
    tags <- setdiff(rownames(maf), unique(unlist(allTar)))

    maf[tags, 'CaTag'] <- paste0(maf[tags, 'CaTag'], 'R')
  }

  return(maf)
}


## helper function to target variants in specific region
mutRegionTag <- function(mutLoc, bedSingle){
  # count <- bedSingle[1, 'Num']
  inRegion <- rep(NA, nrow(mutLoc))
  for (i in seq_len(nrow(mutLoc))) {
    if (mutLoc[i, 'Start_Position'] >= bedSingle[1, 'chromStart'] &
        mutLoc[i, 'End_Position'] <= bedSingle[1, 'chromEnd']) {
      # count <- count  + 1
      inRegion[i] <- rownames(mutLoc[i, ])
    }
  }
  return(na.omit(inRegion))
}
