#' mutFilterQual
#' @description Filter variants in low sequencing quality or low confidence.
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' @param tumorDP Threshold of tumor total depth. Default: 20
#' @param normalDP Threshold of normal total depth. Default: 10
#' @param tumorAD Threshold of tumor alternative allele depth. Default: 10
#' @param VAF Threshold of VAF value. Default: 0.05
#' @param VAFratio Threshold of VAF ratio (tVAF/nVAF). Default: 5
#' 
#' @return An MAF data frame after filtration for sequencing quality
#' 
#' @export mutFilterQual

mutFilterQual <- function(maf, tumorDP = 20, normalDP = 10, 
                          tumorAD = 10, VAF = 0.05, VAFratio = 5) {
  
  ## VAF filtering
  maf_filtered <- as.data.frame(maf[which(maf$VAF > VAF), ])
  if (nrow(maf_filtered) == 0) {
    return(maf_filtered)
  }
  
  ## tumorAD, tumorDP, normalDP, VAF ratio filtering
  maf_filtered <- maf_filtered[which((maf_filtered$t_alt_count >= tumorAD) & 
                                       (maf_filtered$t_depth >= tumorDP) &
                                       (maf_filtered$n_ref_count >= normalDP)),]
  VAFr <- (maf_filtered$t_alt_count/maf_filtered$t_depth)/
    (maf_filtered$n_alt_count/maf_filtered$n_depth)
  maf_filtered <- maf_filtered[VAFr >= VAFratio, ]
  
  
  if (nrow(maf_filtered) == 0){
    message('No mutation left after sequencing quality filtration.')
  }else{
    rownames(maf_filtered) <- seq_len(nrow(maf_filtered))
  }
  return(maf_filtered)
}

