#' mutFilterAdj
#' @description Filter variants based on adjacent indel tags
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' @param maxIndelLen Maximum length of indel accepted to be included.
#' Default: 50
#' @param minInterval Minimum length of interval between an SNV and an indel
#' accepted to be included. Default: 10
#'
#' @return An MAF data frame after filtration for adjacent variants.
#'
#' @import dplyr
#'
#' @export mutFilterAdj
#' @examples
#' maf <- vcfToMAF(system.file("extdata", "GC48-2_mutect2.vep.vcf",
#' package = "CaMutQC"))
#' mafF <- mutFilterAdj(maf)


mutFilterAdj <- function(maf, maxIndelLen = 50, minInterval = 10){

  # sort and order variants in maf file
  mafFile <- arrange(maf, Chromosome, Start_Position)
  # mafFile$Start_Position <- as.numeric(mafFile$Start_Position)
  # mafFile$End_Position <- as.numeric(mafFile$End_Position)

  # calculate the distance between variants
  mafFile <- mafFile %>% group_by(Chromosome) %>%
    mutate(pre_diff = Start_Position - lag(End_Position,
                                           default = Start_Position[1]),
           next_diff = lead(Start_Position, 
                            default = Start_Position[1]) - End_Position,
           prev_mut_length = lag(nchar(Reference_Allele), default = 1),
           next_mut_length = lead(nchar(Reference_Allele), default = 1),
           prev_mut_type = lag(Variant_Type, default = "SNP"), 
           next_mut_type = lead(Variant_Type, default = "SNP"))
  mafFile <- data.frame(mafFile)
  
  # either of the conditions should be satisfied to avoid being filtered
  ## 1. the current mutation be analyzed is an INDEL
  ## 2. the current mutation (SNP) is far enough from the previous mutation if
  ##    the previous mutation is an indel within max length, and, 
  ##    far enough from the next mutation if the next mutation is an indel 
  ##    within max length
  condition1 <- (mafFile$Variant_Type == "INS" | mafFile$Variant_Type == "DEL")
  condition2_1 <- (mafFile$pre_diff >= minInterval | 
                     mafFile$prev_mut_length >= maxIndelLen | 
                     mafFile$prev_mut_type == "SNP" | mafFile$pre_diff < 0)
  condition2_2 <- (mafFile$next_diff >= minInterval | 
                     mafFile$next_mut_length >= maxIndelLen | 
                     mafFile$next_mut_type == "SNP" | mafFile$next_diff < 0)
  
  n_tags <- setdiff(rownames(mafFile), 
                    rownames(mafFile[((condition2_1 & condition2_2) | 
                                        condition1),]))
  
  mafFile[n_tags, 'CaTag'] <- paste0(mafFile[n_tags, 'CaTag'] , 'A')

  return(subset(mafFile, select = c(-pre_diff, -next_diff, -prev_mut_length,
                                    -next_mut_length, -prev_mut_type, 
                                    -next_mut_type)))
}

