#' mutFilterPON
#' @description Filter variants in PON
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' 
#' @return An MAF data frame after PON filtering.
#' @import vcfR, stringr


## PON filtering using external dataset and info flag
mutFilterPON <- function(maf) {
  
  # first filter using the flag in maf column
  if('PON' %in% colnames(maf)) {
    maf <- data.frame(maf[is.na(maf$PON), ])
  }
  rownames(maf) <- 1:nrow(maf)
  
  discard <- c()
  
  # joint useful information for matching
  if (unique(maf$NCBI_Build) == 'GRCh37') {
    somatic37 <- read.vcfR('../inst/extdata/somatic-b37_Mutect2-exome-panel.vcf')
    ext37 <- rep(NA, nrow(somatic37@fix))
    for (k in 1:nrow(somatic37@fix)) {
      ext37[k] <- paste(somatic37@fix[k, c(1, 2, 4, 5)], collapse = ";")
    }
    
    # combine useful info for matching
    for (i in 1:nrow(maf)) {
      info <- paste(c(str_extract(maf$Chromosome[i], '[:digit:]+'), 
                      maf[i, c(6, 11, 13)]), collapse = ";")
      if (info %in% ext37) {
        discard <- c(discard, i)
      }
    }
  } else if (unique(maf$NCBI_Build) == 'GRCh38') {
    somatic38 <- read.vcfR('../inst/extdata/somatic-hg38_1000g_pon.hg38.vcf')
    ext38 <- rep(NA, nrow(somatic38@fix))
    for (k in 1:nrow(somatic38@fix)) {
      ext38[k] <- paste(somatic38@fix[k, c(1, 2, 4, 5)], collapse = ";")
    }
    
    # combine useful info for matching
    for (i in 1:nrow(maf)) {
      info <- paste(c(str_extract(maf$Chromosome[i], '[:digit:]+'), 
                      maf[i, c(6, 11, 13)]), collapse = ";")
      if (info %in% ext38) {
        discard <- c(discard, i)
      }
    }
    
  } else if (!(unique(maf$NCBI_Build) %in% c('GRCh38', 'GRCh37'))){
    stop('Invaild genome version.')
  }else if (length(unique(maf$NCBI_Build)) > 1){
    stop('There are more than one version of genome in this dataset.')
  }
  
  # discard somatic mutation
  if (is.null(discard)){
    return(maf)
  }else {
    maf_filtered <- as.data.frame(maf[-discard, ])
    if (nrow(maf_filtered) == 0){
      message('No mutation left after PON filtering.')
    }else{
      rownames(maf_filtered) <- 1:nrow(maf_filtered)
    }
    return(maf_filtered)
  }
}
