#' mutFilterAdj
#' @description Filter alleles based on adjacent indels/tags
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' @param maxindelLen Maximum length of indel accepted to be included. 
#' Default: 50
#' @param maxInterval Maximum length of interval between an SNV and an indel 
#' accepted to be included. Default: 10
#' 
#' @return An MAF data frame after filtration for adjacent alleles.
#' 
#' @import dplyr
#' 
#' @export mutFilterAdj

mutFilterAdj <- function(maf, maxindelLen = 50, maxInterval = 10){
  
  # sort and order variants in maf file
  mafFile <- arrange(maf, Chromosome, Start_Position)
  mafFile$Start_Position <- as.numeric(mafFile$Start_Position)
  mafFile$End_Position <- as.numeric(mafFile$End_Position)
  
  # calculate the distance between variants
  mafFile <- mafFile %>% group_by(Chromosome) %>% 
    mutate(pre_diff = Start_Position - lag(Start_Position, 
                                           default = Start_Position[1]), 
           next_diff = End_Position - lag(End_Position, 
                                          default = End_Position[1]))
  
  # filter variants
  maf_filtered <- filter(mafFile, nchar(Reference_Allele) < maxindelLen & 
                        (pre_diff == 0 & next_diff == 0) |
                        (pre_diff >= maxInterval & next_diff >= maxInterval))
  
  # return results
  if (nrow(maf_filtered) == 0){
    message('No mutation left after Adjacent indel filtration.')
  }else{
    maf_filtered <- as.data.frame(maf_filtered)
    rownames(maf_filtered) <- seq_len(nrow(maf_filtered))
  }
  return(maf_filtered[, c(-116, -117)])
}




