#'
#' @description Filter variants of certain types
#'
#' @param rawmaf An MAF object, generated by \code{\link{vcfToMaf}} function
#' @param vcf_add A data frame contains the columns include and after 
#' the FORMAT column, come from the gt slot of a vcfR-class object
#' @param CSQ_info A data frame contains CSQ information extracted from VCF file 
#' and matches corresponding lines in MAF file, 
#' generated by \code{\link{getCSQ}} function
#' @param types A group of variant classifications will be filtered, 
#' including 'nonExomic' and 'synonymous'. Default: 'nonExomic'. 
#' 
#'

mutFilterType <- function(rawmaf, vcf_add, CSQ_info, types = 'nonExomic') {
  
  # process types
  if (types == 'nonExomic') {
    types <- c('RNA', 'Intron', 'IGR', '5\'Flank', '3\'Flank', 
               '5\'UTR', '3\'UTR')
  } else if (types == 'synonymous') {
    types <- c("3'UTR", "5'UTR", "3'Flank", "Targeted_Region", "Silent", 
               "Intron", "RNA", "IGR", "Splice_Region", "5'Flank", "lincRNA", 
               "De_novo_Start_InFrame", "De_novo_Start_OutOfFrame", 
               "Start_Codon_Ins", "Start_Codon_SNP", "Stop_Codon_Del")
  } else {
    stop('Please select one group of types from nonExomic and synonymous.')
  }
  
  maf <- data.frame(rawmaf[which(!(rawmaf$Variant_Classification %in% types)), ])
  CSQ_f <- data.frame(CSQ_info[which(!(rawmaf$Variant_Classification %in% types)), ])
  vcf <- data.frame(vcf_add[which(!(rawmaf$Variant_Classification %in% types)), ])
  
  if (nrow(maf) == 0) {
    message('Note: no mutation left after variant type filtering')
    return(list(maf, vcf, CSQ_f))
  } else {
    rownames(maf) <- 1:nrow(maf)
    rownames(CSQ_f) <- 1:nrow(CSQ_f)
    rownames(vcf) <- 1:nrow(vcf)
    return(list(maf, vcf, CSQ_f))
  }
}
