#' mutFilterQual
#' @description Filter variants in low sequencing quality or low confidence.
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' @param tumorSampleName Label of the tumor sample, should be one of the 
#' column names of maf. If it is set as 'Extracted', tumorSampleName would be
#' extracted automatically from the maf data frame. Default: 'Extracted'.
#' @param normalSampleName Label of the normal sample, should be one of the 
#' column names of maf. If it is set as 'Extracted', normalSampleName would be
#' extracted automatically from the maf data frame. Default: 'Extracted'.
#' @param tumorDP Threshold of tumor total depth. Default: 20
#' @param normalDP Threshold of normal total depth. Default: 10
#' @param tumorAD Threshold of tumor alternative allele depth. Default: 10
#' @param VAF Threshold of VAF value. Default: 0.05
#' @param VAFratio Threshold of VAF ratio (tVAF/nVAF). Default: 5
#' 
#' @return An MAF data frame after filtering for sequencing quality

mutFilterQual <- function(maf, tumorSampleName = 'Extracted', 
                          normalSampleName = 'Extracted', tumorDP = 20, 
                          normalDP = 10, tumorAD = 10, VAF = 0.05, 
                          VAFratio = 5) {
  
  ## VAF filtering
  maf_filtered <- as.data.frame(maf[which(maf$VAF > VAF), ])
  if (nrow(maf_filtered) == 0) {
    return(maf_filtered)
  }
  
  ## get tumorSampleName and normalSampleName
  if (tumorSampleName == 'Extracted'){
    tumorSampleName <- unique(maf_filtered$Tumor_Sample_Barcode)
  } else if(!(tumorSampleName %in% colnames(maf_filtered))) {
    stop('Invaild tumorSampleName.')
  }
  
  if (normalSampleName == 'Extracted'){
    normalSampleName <- unique(maf_filtered$Matched_Norm_Sample_Barcode)
  }else if(!(normalSampleName %in% colnames(maf_filtered))) {
    stop('Invaild normalSampleName.')
  }
  
  ## tumorAD, tumorDP, normalDP, VAF ratio filtering
  discard <- c()
  for (n in 1:nrow(maf_filtered)) {
    DP_loc <- strsplit(maf_filtered[n, 'FORMAT'], ":")[[1]] == 'DP'
    AD_loc <- strsplit(maf_filtered[n, 'FORMAT'], ":")[[1]] == 'AD'
    tAD <- as.numeric(strsplit(strsplit(maf_filtered[n, tumorSampleName], 
                                        ":")[[1]][AD_loc], ",")[[1]][2])
    nAD <- as.numeric(strsplit(strsplit(maf_filtered[n, normalSampleName], 
                                        ":")[[1]][AD_loc], ",")[[1]][2])
    tDP <- as.numeric(strsplit(maf_filtered[n, tumorSampleName], 
                               ":")[[1]][DP_loc])
    nDP <- as.numeric(strsplit(maf_filtered[n, normalSampleName], 
                               ":")[[1]][DP_loc])
    
    if ((tAD < tumorAD) | (tDP < tumorDP) | (nDP < normalDP)) {
      discard <- c(discard, n)
    } else if (((tAD/tDP)/ (nAD/nDP)) < VAFratio) {
      discard <- c(discard, n)
    }
  }
  
  if (is.null(discard)){
    return(maf_filtered)
  }else {
    maf_filtered <- as.data.frame(maf_filtered[-discard, ])
    if (nrow(maf_filtered) == 0){
      message('No mutation left after sequencing quality filtering.')
    }else{
      rownames(maf_filtered) <- 1:nrow(maf_filtered)
    }
    return(maf_filtered)
  }
  
}

