#' mutFilterTech
#' @description Filter potential artifacts produced through technical issue, 
#' including filtration for sequencing quality, strand bias, adjacent indel 
#' tag and FILTER field.
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' @param tumorDP Threshold of tumor total depth. Default: 20
#' @param normalDP Threshold of normal total depth. Default: 10
#' @param tumorAD Threshold of tumor alternative allele depth. Default: 10
#' @param VAF Threshold of VAF value. Default: 0.05
#' @param VAFratio Threshold of VAF ratio (tVAF/nVAF). Default: 5
#' @param tumorSampleName Label of the tumor sample, should be one of the 
#' column names of maf. If it is set as 'Extracted', tumorSampleName would be
#' extracted automatically from the maf data frame. Default: 'Extracted'.
#' @param SBmethod Method will be used to detect strand bias, 
#' including 'SOR' and 'Fisher'. Default: 'SOR'. SOR: StrandOddsRatio 
#' (https://gatk.broadinstitute.org/hc/en-us/articles/360041849111-
#' StrandOddsRatio)
#' @param SBscore Cutoff strand bias score used to filter variants.
#' Default: 3
#' @param maxindelLen Maximum length of indel accepted to be included. 
#' Default: 50
#' @param maxInterval Maximum length of interval between an SNV and an indel 
#' accepted to be included. Default: 10
#' @param tagFILTER Variants with spcific tag in the FILTER column will be kept,
#' Default: 'PASS'
#' 
#' @import dplyr
#' 
#' @return An MAF data frame after filtration for technical issue
#' 
#' @export mutFilterTech


mutFilterTech <- function(maf, tumorDP = 20, normalDP = 10, tumorAD = 10, 
                          VAF = 0.05, VAFratio = 5, tumorSampleName = 'Extracted', 
                          SBmethod = 'SOR', SBscore = 3, maxindelLen = 50, 
                          maxInterval = 10, tagFILTER = 'PASS'){
  
  # sequencing quality filtration
  message('Filtration for sequencing quality is in process.')
  maf <- mutFilterQual(maf, tumorDP = tumorDP, tumorAD = tumorAD,
                       normalDP = normalDP, VAF = VAF, VAFratio = VAFratio)
  
  # strand bias filtration
  message('Filtration for strand bias is in process.')
  maf <- mutFilterSB(maf, tumorSampleName = tumorSampleName, method = SBmethod,
                     SBscore = SBscore)
  
  # adjacent indel tag filtration
  message('Filtration for adjacent indel is in process.')
  maf <- mutFilterAdj(maf, maxindelLen = maxindelLen, maxInterval = maxInterval)
  
  # FILTER field filtration
  message('Filtration for FILTER column is in process.')
  tagFilter <- rownames(maf[(!(is.na(maf$FILTER)) & 
                               (maf$FILTER != tagFILTER)), ])
  maf[tagFilter, 'CaTag'] <-  paste0(maf[tagFilter, 'CaTag'], 'F')
  
  return(maf)
}

