#'
#' @description Filter SNP variants based on VAF and clinical significance
#'
#' @param rawmaf An MAF object, generated by \code{\link{vcfToMaf}} function
#' @param vcf_add A data frame contains the columns include and after 
#' the FORMAT column, come from the gt slot of a vcfR-class object
#' @param CSQ_info A data frame contains CSQ information extracted from VCF file 
#' and matches corresponding lines in MAF file, 
#' generated by \code{\link{getCSQ}} function
#' @param VAFcutoff Threshold of VAF value, variants with VAF lower than cutoff
#' will be filtered. Default: 0.05
#' 
#'

mutFilterSNP <- function(rawmaf, vcf_add, CSQ_info, VAFcutoff = 0.05) {
  discard <- c()
  for (i in 1:nrow(rawmaf)) {
    if (rawmaf[i, 'VAF'] < VAFcutoff) {
      if (length(grep('athogenic', CSQ_info[i, 'CLIN_SIG'])) == 0){
        discard <- c(discard, i)
      }
    }
  }
  
  if (is.null(discard)){
    message('Note: no mutation filtered for SNP.')
    return(list(rawmaf, vcf_add, CSQ_info))
  }else {
    maf <- rawmaf[-discard, ]
    vcf <- vcf_add[-discard, ]
    CSQ_f <- CSQ_info[-discard, ]
  }
  
  if (nrow(maf) == 0) {
    
    message('Note: no mutation left after SNP filtering')
    return(list(maf, vcf, CSQ_f))
  } else {
    rownames(maf) <- 1:nrow(maf)
    rownames(CSQ_f) <- 1:nrow(CSQ_f)
    rownames(vcf) <- 1:nrow(vcf)
    return(list(maf, vcf, CSQ_f))
  }
}
