#' mutFilterDB
#' @description Filter variants in low sequencing quality or low confidence.
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' @param VAF Threshold of VAF value. Default: 0.01
#' @param ExAc Whether to filter variants listed in ExAc with VAF higher than
#' cutoff(set in VAF parameter). Default: TRUE.
#' @param Genomesprojects1000 Whether to filter variants listed in 
#' Genomesprojects1000 with VAF higher than cutoff(set in VAF parameter). 
#' Default: TRUE.
#' @param ESP6500 Whether to filter variants listed in ESP6500 with VAF higher 
#' than cutoff(set in VAF parameter). Default: TRUE.
#' @param gnomAD Whether to filter variants listed in gnomAD with VAF higher 
#' than cutoff(set in VAF parameter). Default: TRUE.
#' @param COSMIConly Whether to only keep variants in COSMIC. Default: FALSE.
#' 
#' @return An MAF data frame after filtrationfor sequencing quality
#' 
#' @export mutFilterDB

mutFilterDB <- function(maf, VAF = 0.01, ExAc = TRUE, Genomesprojects1000 = TRUE,
                        ESP6500 = TRUE, gnomAD = TRUE, COSMIConly = FALSE){
  
  # ExAc filtration
  if (ExAc){
    if ('AC' %in% colnames(maf)){
      maf <- maf[((is.na(maf$AC)) | (maf$AC == 0) | (maf$AC == '')), ]
    }else{
      message(paste0('ExAc information cannot be found in VCF file.' ,
      ' No variants will be filtered based on ExAc.'))
    } 
  }
  
  # 1000 Genomesprojects filtration
  if (Genomesprojects1000){
    if ('AF' %in% colnames(maf)){
      maf <- maf[((as.numeric(maf$AF) < VAF) | (maf$AF == '')), ]
    }else{
      message(paste0('1000 Genomesprojects information cannot be',
                     ' found in VCF file. No variants will be filtered based' ,
                     ' on 1000 Genomesprojects.'))
    }
  }
  
  # gnomAD filtration
  if (gnomAD){
    if ('gnomAD_AF' %in% colnames(maf)){
      maf <- maf[((as.numeric(maf$gnomAD_AF) < VAF) | (maf$gnomAD_AF == '')), ]
    }else{
      message(paste0('gnomAD information cannot be found in VCF file.' ,
                     ' No variants will be filtered based on gnomAD.'))
    }
  }
  
  # ESP6500 filtration
  if (ESP6500){
    if ('AA_AF' %in% colnames(maf)){
      maf <- maf[((as.numeric(maf$AA_AF) < VAF) | (maf$AA_AF == '')), ]
    }
    if ('EA_AF' %in% colnames(maf)){
      maf <- maf[((as.numeric(maf$EA_AF) < VAF) | (maf$EA_AF == '')), ]
    }else{
      message(paste0('ESP6500 information cannot be found in VCF file.' ,
                     ' No variants will be filtered based on ESP6500.'))
    }
  }
  
  if (COSMIConly) {
    maf <- maf[grep('COS', maf[, 'Existing_variation']), ]
  }
  
  return(maf)
}



