#' mutFilterTech
#' @description Filter potential artifacts produced through technical issue,
#' including filtration for sequencing quality, strand bias, adjacent indel
#' tag and FILTER field.
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' @param panel The sequencing panel applied on the dataset. Parameters
#' for \code{\link{mutFilterQual}} function are set differently for different
#' panels. Default: "Customized". Options: "MSKCC", "WES".
#' @param tumorDP Threshold of tumor total depth. Default: 0
#' @param normalDP Threshold of normal total depth. Default: 0
#' @param tumorAD Threshold of tumor alternative allele depth. Default: 0
#' @param normalAD Threshold of normal alternative allele depth. Default: Inf
#' @param VAF Threshold of VAF value. Default: 0
#' @param VAFratio Threshold of VAF ratio (tVAF/nVAF). Default: 0
#' @param SBmethod Method will be used to detect strand bias,
#' including 'SOR' and 'Fisher'. Default: 'SOR'. SOR: StrandOddsRatio
#' (https://gatk.broadinstitute.org/hc/en-us/articles/360041849111-
#' StrandOddsRatio)
#' @param SBscore Cutoff strand bias score used to filter variants.
#' Default: 3
#' @param maxIndelLen Maximum length of indel accepted to be included.
#' Default: 50
#' @param minInterval Minimum length of interval between an SNV and an indel
#' accepted to be included. Default: 10
#' @param tagFILTER Variants with specific tag in the FILTER column will be kept,
#' set to NULL if you want to skip this filter. Default: 'PASS'
#'
#' @import dplyr
#'
#' @return An MAF data frame after filtration for technical issue
#'
#' @export mutFilterTech
#' @examples
#' maf <- vcfToMAF(system.file("extdata", "GC48-2_mutect2.vep.vcf",
#' package = "CaMutQC"))
#' mafF <- mutFilterTech(maf)

mutFilterTech <- function(maf, panel = "Customized", tumorDP = 0, normalDP = 0,
                          tumorAD = 0, normalAD = Inf, VAF = 0, VAFratio = 0,
                          SBmethod = 'SOR', SBscore = 3, maxIndelLen = 50,
                          minInterval = 10, tagFILTER = 'PASS'){

  # sequencing quality filtration
  message('Filtration for sequencing quality is in process.')
  maf <- mutFilterQual(maf, panel = panel, tumorDP = tumorDP, tumorAD = tumorAD,
                       normalDP = normalDP, normalAD = normalAD,
                       VAF = VAF, VAFratio = VAFratio)

  # strand bias filtration
  message('Filtration for strand bias is in process.')
  maf <- mutFilterSB(maf, method = SBmethod, SBscore = SBscore)

  # adjacent indel tag filtration
  message('Filtration for adjacent indel is in process.')
  maf <- mutFilterAdj(maf, maxIndelLen = maxIndelLen, minInterval = minInterval)

  # FILTER field filtration
  if (!(is.null(tagFILTER))){
    message('Filtration for FILTER column is in process.')
    tagFilter <- rownames(maf[(!(is.na(maf$FILTER)) &
                                 (maf$FILTER != tagFILTER)), ])
    maf[tagFilter, 'CaTag'] <-  paste0(maf[tagFilter, 'CaTag'], 'F')
  }

  # complete filtration
  message('Filtration for technical issue is done!')
  return(maf)
}

