#' mutFilterPON
#' @description Filter variants in Panel of Normals
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' 
#' @return An MAF data frame after PON filtration.
#' @import vcfR, stringr
#' 
#' @export mutFilterPON


## PON filtration using external dataset and info flag
mutFilterPON <- function(maf) {
  
  # first filter using the flag in maf column
  if('PON' %in% colnames(maf)) {
    maf <- data.frame(maf[is.na(maf$PON), ])
  }
  rownames(maf) <- seq_len(nrow(maf))
  
  discard <- c()
  
  # joint useful information for matching
  if (unique(maf$NCBI_Build) == 'GRCh37') {
    message('Loading PON data...')
    invisible(capture.output(somatic37 <- read.vcfR(
      '../inst/extdata/somatic-b37_Mutect2-exome-panel.vcf')))
    message('PON data has been loaded successfully!')
    ext37 <- rep(NA, nrow(somatic37@fix))
    for (k in seq_len(nrow(somatic37@fix))) {
      ext37[k] <- paste(somatic37@fix[k, c(1, 2, 4, 5)], collapse = ";")
    }
    
    # combine useful info for matching
    for (i in seq_len(nrow(maf))) {
      info <- paste(c(str_extract(maf$Chromosome[i], '[:digit:]+'), 
                      maf[i, c(6, 11, 13)]), collapse = ";")
      if (info %in% ext37) {
        discard <- c(discard, i)
      }
    }
  } else if (unique(maf$NCBI_Build) == 'GRCh38') {
    message('Loading PON data...')
    invisible(capture.output(somatic38 <- read.vcfR(
      '../inst/extdata/somatic-hg38_1000g_pon.hg38.vcf')))
    message('PON data has been loaded successfully!')
        ext38 <- rep(NA, nrow(somatic38@fix))
    for (k in seq_len(nrow(somatic38@fix))) {
      ext38[k] <- paste(somatic38@fix[k, c(1, 2, 4, 5)], collapse = ";")
    }
    
    # combine useful info for matching
    for (i in seq_len(nrow(maf))) {
      info <- paste(c(str_extract(maf$Chromosome[i], '[:digit:]+'), 
                      maf[i, c(6, 11, 13)]), collapse = ";")
      if (info %in% ext38) {
        discard <- c(discard, i)
      }
    }
    
  } else if (!(unique(maf$NCBI_Build) %in% c('GRCh38', 'GRCh37'))){
    stop('Invaild genome version.')
  }else if (length(unique(maf$NCBI_Build)) > 1){
    stop('There are more than one version of genome in this dataset.')
  }
  
  # add tag
  maf[discard, 'CaTag'] <- paste0(maf[discard, 'CaTag'], 'P')
  return(maf)
}
