#' mutFilterAdj
#' @description Filter alleles based on adjacent indel tags
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' @param maxIndelLen Maximum length of indel accepted to be included.
#' Default: 50.
#' @param minInterval Minimum length of interval between an SNV and an indel
#' accepted to be included. Default: 10.
#'
#' @return An MAF data frame after filtration for adjacent alleles.
#'
#' @import dplyr
#'
#' @export mutFilterAdj
#' @examples
#' maf <- vcfToMAF(system.file("extdata", "GC48-2_mutect2.vep.vcf",
#' package = "CaMutQC"))
#' mafF <- mutFilterAdj(maf)


mutFilterAdj <- function(maf, maxIndelLen = 50, minInterval = 10){

  # sort and order variants in maf file
  mafProcess <- arrange(maf, Chromosome, Start_Position)
  # mafProcess$Start_Position <- as.numeric(mafProcess$Start_Position)
  # mafProcess$End_Position <- as.numeric(mafProcess$End_Position)

  # calculate the distance between variants
  mafProcess <- mafProcess %>% group_by(Chromosome) %>%
    mutate(pre_diff = Start_Position - lag(Start_Position,
                                           default = Start_Position[1]),
           next_diff = End_Position - lag(End_Position,
                                          default = End_Position[1]))

  # store indexes of mutations that pass the filter
  n_tags <- rownames(mafProcess[(nchar(mafProcess$Reference_Allele)
                                    <= maxIndelLen) &
                        ((mafProcess$pre_diff == 0 & mafProcess$next_diff == 0) |
                        (mafProcess$pre_diff >= minInterval &
                           mafProcess$next_diff >= minInterval)), ])
  mafProcess <- data.frame(mafProcess)

  mafProcess[!(rownames(mafProcess) %in% n_tags), 'CaTag'] <-
    paste0(mafProcess[!(rownames(mafProcess) %in% n_tags), 'CaTag'] , 'A')

  return(subset(mafProcess, select = c(-pre_diff, -next_diff)))
}

