#' calTMB
#' @description Calculate the Tumor Mutational Burden (TMB) in specific regions.
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' @param bedFile A file in bed format that contains region information.
#' @param tumorSampleName Label of the tumor sample, should be one of the 
#' column names of maf.
#' @param assay Methodology and assay will be applied as a reference, including
#' 'MSKCC-v3', 'MSKCC-v2', 'MSKCC-v1', 'FoundationOne' and 'Pan-Cancer-Panel'.
#' Default: 'MSKCC-v3'.
#' @param VAF VAF threshold used for filtering variants. Default: 0.05.
#' @param AD Allele read count threshold used for filtering. Default: 10.
#' @param mutType A group of variant classifications that will be kept, 
#' including 'exomic' and 'nonsynonymous'. Default: 'nonsynonymous'. 
#' @param bedFilter Whether to filter the information in bed file or not. 
#' Default: TRUE
#' 
#' @import stringr
#' @return A TMB value.


calTMB <- function(maf, bedFile, tumorSampleName, assay = 'MSKCC-v3', 
                   VAF = 0.05, AD = 5, mutType = 'nonsynonymous', 
                   bedFilter = TRUE) {
  
  # bed <- read.table('../Data/test-bed')
  bed <- read.table(bedFile)
  chromVaild <- c(paste0('chr', 1:22), 'chrX', 'chrY')
  
  ## check for validation of bed file
  if (nrow(bed) < 3 | any(substring(bed$V1, 1,3) != 'chr') | 
      typeof(bed$V2) != 'integer' | typeof(bed$V3) != 'integer'){
    stop('Invaild bed file. Please input vaild bed file with at least 3 columns of porper type.')
  }
  
  if(any(!(bed$V1 %in% chromVaild))){
    if (!bedFilter){
      stop('Invaild chrom in bed File. Please turn on bedFiler if you want to filter the bed file')
    } else {
      bed <- bed[which(bed$V1 %in% chromVaild), ]
    }
  }
  
  ## select variants based on chosen assay
  if (assay == 'MSKCC-v3') {
    mskcc_v3 <- read.table('../inst/extdata/MSK-IMPACT_gene_v3_468.txt')
    maf <- maf[which(maf$Hugo_Symbol %in% mskcc_v3$V1), ]
  }else if (assay == 'MSKCC-v2') {
    mskcc_v2 <- read.table('../inst/extdata/MSK-IMPACT_gene_v2_410.txt')
    maf <- maf[which(maf$Hugo_Symbol %in% mskcc_v2$V1), ]
  }else if (assay == 'MSKCC-v2') {
    mskcc_v1 <- read.table('../inst/extdata/MSK-IMPACT_gene_v1_341.txt')
    maf <- maf[which(maf$Hugo_Symbol %in% mskcc_v1$V1), ]
  }else if (assay == 'FoundationOne') {
    FoundationOne <- read.table('../inst/extdata/FoundationOne_genelist.txt')
    maf <- maf[which(maf$Hugo_Symbol %in% FoundationOne$V1), ]
  }else if (assay == 'Pan-Cancer-Panel') {
    pan <- read.table('../inst/extdata/TMB_panel_genelist.txt', header = TRUE)
    maf <- maf[which(maf$Hugo_Symbol %in% pan$Gene_Symbol), ]
  }else{
    stop("Invalid assay. Please select from \'MSKCC-v3\', \'MSKCC-v2\', 
    \'MSKCC-v1\', \'FoundationOne\' and \'Pan-Cancer-Panel\'. ")
  }
  
  ## select certain variant type
  if (mutType == 'exomic') {
    mutType <- c('RNA', 'Intron', 'IGR', '5\'Flank', '3\'Flank', 
               '5\'UTR', '3\'UTR')
  } else if (mutType == 'nonsynonymous') {
    mutType <- c("3'UTR", "5'UTR", "3'Flank", "Targeted_Region", "Silent", 
               "Intron", "RNA", "IGR", "Splice_Region", "5'Flank", "lincRNA", 
               "De_novo_Start_InFrame", "De_novo_Start_OutOfFrame", 
               "Start_Codon_Ins", "Start_Codon_SNP", "Stop_Codon_Del")
  } else {
    stop('Please select one group of types from exomic and nonsynonymous.')
  }
  
  maf <- maf[which(!(maf$Variant_Classification %in% mutType)), ]
  
  ## VAF and AD filtering (using mutFilterQual)
  maf <- mutFilterQual(maf, tumorDP = 0, normalDP = 0, 
                       tumorAD = AD, VAF = VAF, VAFratio = 0)
  
  ## sort bed object
  bedProc <- bed[, 1:3]
  colnames(bedProc) <- c('chrom', 'chromStart', 'chromEnd')
  bedProc <- unique(cbind(bedProc, interval = bedProc$chromEnd - bedProc$chromStart, 
                   Num = 0))
  rownames(bedProc) <- 1:nrow(bedProc)
  ## count mutation
  # maf <- arrange(maf, Chromosome, Start_Position)
  # group_by(Chromosome)
  
  # bedSorted <- arrange(bedProc, chrom, chromStart)
  maf <- maf[, c("Chromosome", "Start_Position", "End_Position")]
  chrs <- unique(maf$Chromosome)
  # nums <- rep(0, length(chrs))
  for (c in  1:length(chrs)) {
    maf_target <- maf[which(maf$Chromosome == chrs[c]), ]
    bed_target <- bedProc[which(bedProc$chrom == chrs[c]), ]
    l <- rep(0, nrow(bed_target))
    for(i in 1:nrow(bed_target)) {
      l[i] <- mutCountRegion(maf_target, bed_target[i, ])
    }
    bedProc[which(bedProc$chrom == chrs[c]), 'Num'] <- l
  }
  
  TMB <- mean(bedProc$Num/bedProc$interval) * 1000000
  return(TMB)
}


## helper function for counting variants in specific region
mutCountRegion <- function(mutLoc, bedSingle){
  count <- bedSingle[1, 'Num']
  for (i in 1:nrow(mutLoc)) {
    if (mutLoc[i, 'Start_Position'] >= bedSingle[1, 'chromStart'] & 
        mutLoc[i, 'End_Position'] <= bedSingle[1, 'chromEnd']) {
      count <- count  + 1
    }
  }
  return(count)
}

