#' calTMB
#' @description Calculate Tumor Mutational Burden (TMB) in specific regions.
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' @param bedFile A file in bed format that contains region information.
#' @param assay Methodology and assay will be applied as a reference, including
#' 'MSKCC-v3', 'MSKCC-v2', 'MSKCC-v1', 'FoundationOne', 'Pan-Cancer Panel' and
#' 'Customized'. Default: 'MSKCC-v3'.
#' @param genelist A vector of panel gene list, only useful when assay is set to
#' 'Customized'.
#' @param VAF VAF threshold used for filtering variants. Default: 0.05.
#' @param AD Allele read count threshold used for filtering. Default: 10.
#' @param mutType A group of variant classifications that will be kept, 
#' only useful when assay is set to 'Pan-Cancer Panel' or 'Customized',
#' including 'exonic' and 'nonsynonymous'. Default: 'nonsynonymous'. 
#' @param bedFilter Whether to filter the information in bed file or not, which 
#' only leaves segments in Chr1-Ch22, ChrX and ChrY. Default: TRUE
#' 
#' @return A TMB value.
#' 
#' @export calTMB


calTMB <- function(maf, bedFile, assay = 'MSKCC-v3', genelist = NULL, 
                   VAF = 0.05, AD = 5, mutType = 'nonsynonymous', 
                   bedFilter = TRUE) {
  
  # bed <- read.table('../Data/test-bed')
  bed <- read.table(bedFile)
  chromVaild <- c(paste0('chr', 1:22), 'chrX', 'chrY')
  
  ## check for validation of bed file
  if (nrow(bed) < 3 | any(substring(bed$V1, 1,3) != 'chr') | 
      typeof(bed$V2) != 'integer' | typeof(bed$V3) != 'integer'){
    stop(paste0('Invaild bed file. Please input vaild bed file',
                ' with at least 3 columns of porper type.'))  }
  
  if(any(!(bed$V1 %in% chromVaild))){
    if (!bedFilter){
      stop(paste0('Invaild chrom in bed File. Please turn on bedFiler',
                  ' if you want to filter the bed file'))
    } else {
      bed <- bed[which(bed$V1 %in% chromVaild), ]
    }
  }
  
  ## select variants based on chosen assay
  if (strsplit(assay, split = '-')[[1]][1] == 'MSKCC'){
    
    # filter based on genelist
    if (assay == 'MSKCC-v3') {
      mskcc_v3 <- read.table('../inst/extdata/Panel_gene/MSK-IMPACT_gene_v3_468.txt')
      maf <- maf[which(maf$Hugo_Symbol %in% mskcc_v3$V1), ]
    }else if (assay == 'MSKCC-v2') {
      mskcc_v2 <- read.table('../inst/extdata/Panel_gene/MSK-IMPACT_gene_v2_410.txt')
      maf <- maf[which(maf$Hugo_Symbol %in% mskcc_v2$V1), ]
    }else if (assay == 'MSKCC-v1') {
      mskcc_v1 <- read.table('../inst/extdata/Panel_gene/MSK-IMPACT_gene_v1_341.txt')
      maf <- maf[which(maf$Hugo_Symbol %in% mskcc_v1$V1), ]
    }else{
      stop(paste0('Invalid assay detected.', 
                  ' Please select from \'MSKCC-v3\', \'MSKCC-v2\',', 
                  ' \'MSKCC-v1\', \'FoundationOne\', \'Pan-Cancer Panel\'', 
                  ' and \'Customized\'. '))
    }
    
    # filter on VAF, VAFratio and AD
    maf <- mutFilterQual(maf, tumorAD = 5, VAFratio = 5, VAF = 0.01, 
                         tumorDP = 0, normalDP = 0)
    
    # filter for non-/hotspot genes (COSMIC genes)
    maf_c <- maf[grep('COSV', maf[, 'Existing_variation']), ]
    maf_nonc <- maf[setdiff(rownames(maf_c), rownames(maf)),]
    
    maf_c_f <- mutFilterQual(maf_c, tumorDP = 20, tumorAD = 8, normalDP = 0, 
                             VAF = 0.02, VAFratio = 0)
    maf_nonc_f <- mutFilterQual(maf_nonc, tumorDP = 20, tumorAD = 10, 
                                normalDP = 0, VAF = 0.05, VAFratio = 0)
    maf <- rbind(maf_c_f, maf_nonc_f)
    
    # filter non-exonic variants
    maf <- mutFilterType(maf)
    
  }else if (assay %in% c('FoundationOne', 'Pan-Cancer Panel')) {
    
    # filter noncoding variants
    noncoding <- read.table('../inst/extdata/noncoding.txt', header = TRUE)
    maf <- maf[which(!(maf$One_Consequence %in% 
                         noncoding$Noncoding_Variant_Types)), ]
    
    # filter germline variants(deprecated)
    
    # filter dbsnp variants
    tags1 <- rownames(maf[grep('rs', maf[, 'Existing_variation']), ])
    maf <- maf[setdiff(rownames(maf), tags1), ]
    
    # filter variants with ExAC >= 2
    ## detect whether the VCF file has ExAc annotation or not
    if ('AC' %in% colnames(maf)){
      maf <- maf[maf$AC < 2, ]
    }else{
      warning(paste0('ExAc information cannot be found in VCF file.',
                     ' No variants will be filtered based on ExAc.'))
    }
    
    # filter stop-gain variants in tumor suppressor genes
    TSGs <- read.table('../inst/extdata/TSGenelist.txt', 
                       sep = "\t", header = TRUE)
    TSGs_all <- c(TSGs$GeneSymbol, unique(unlist(strsplit(TSGs$Alias, "\\|"))))
    maf <- maf[(!((maf$Hugo_Symbol %in% TSGs_all) & 
                    (maf$One_Consequence == 'stop_gained'))), ]
    
    # filter variants in COSMIC
    tags2 <- rownames(maf[grep('COSV', maf[, 'Existing_variation']), ])
    maf <- maf[setdiff(rownames(maf), tags2), ]
    
    if (assay == 'FoundationOne') {
      # filter based on genelist
      FoundationOne <- read.table('../inst/extdata/Panel_gene
                                /FoundationOne_genelist.txt')
      maf <- maf[maf$Hugo_Symbol %in% FoundationOne$V1, ]
    }else{
      pan <- read.table('../inst/extdata/Panel_gene/TMB_panel_genelist.txt', 
                        header = TRUE)
      maf <- maf[maf$Hugo_Symbol %in% pan$Gene_Symbol, ]
      maf <- mutFilterQual(maf, tumorDP = 0, normalDP = 0, 
                           tumorAD = 0, VAF = 0.05, VAFratio = 0)
    }
    
  }else if (assay == 'Customized'){
    
    ## genelist
    if (!(is.null(genelist))){
      maf <- maf[(maf$Hugo_Symbol %in% genelist), ]
      
    }
    ## select certain variant type
    maf <- mutFilterType(maf, keepType = mutType)
    
    ## VAF and AD filtering (using mutFilterQual)
    maf <- mutFilterQual(maf, tumorDP = 20, normalDP = 10, 
                         tumorAD = AD, VAF = VAF, VAFratio = 5)
  }else{
    stop(paste0('Invalid assay detected.', 
                ' Please select from \'MSKCC-v3\', \'MSKCC-v2\',', 
                ' \'MSKCC-v1\', \'FoundationOne\', \'Pan-Cancer Panel\'', 
                ' and \'Customized\'. '))
  }
  
  # filter based on CaTag
  maf <- maf[maf$CaTag == '0', ]
  
  # return 0 if no variants left
  if (nrow(maf) == 0){
    return(0)
  }else{
    ## sort bed object
    bedProc <- unique(bed[, 1:3])
    colnames(bedProc) <- c('chrom', 'chromStart', 'chromEnd')
    bedProc <- unique(cbind(bedProc, 
                            interval = bedProc$chromEnd - bedProc$chromStart, 
                            Num = 0))
    rownames(bedProc) <- seq_len(nrow(bedProc))
    ## count mutation
    # maf <- arrange(maf, Chromosome, Start_Position)
    # group_by(Chromosome)
    
    # bedSorted <- arrange(bedProc, chrom, chromStart)
    maf <- maf[, c("Chromosome", "Start_Position", "End_Position")]
    chrs <- unique(maf$Chromosome)
    # nums <- rep(0, length(chrs))
    for (c in seq_len(length(chrs))) {
      maf_target <- maf[which(maf$Chromosome == chrs[c]), ]
      bed_target <- bedProc[which(bedProc$chrom == chrs[c]), ]
      l <- rep(0, nrow(bed_target))
      for(i in seq_len(nrow(bed_target))) {
        l[i] <- mutCountRegion(maf_target, bed_target[i, ])
      }
      bedProc[which(bedProc$chrom == chrs[c]), 'Num'] <- l
    }
    
    TMB <- round(mean(bedProc$Num/bedProc$interval) * 1000000, 3)
    return(TMB)
  }
}


## helper function for counting variants in specific region
mutCountRegion <- function(mutLoc, bedSingle){
  count <- bedSingle[1, 'Num']
  for (i in seq_len(nrow(mutLoc))) {
    if (mutLoc[i, 'Start_Position'] >= bedSingle[1, 'chromStart'] & 
        mutLoc[i, 'End_Position'] <= bedSingle[1, 'chromEnd']) {
      # print(rownames(mutLoc[i, ]))
      count <- count  + 1
    }
  }
  return(count)
}

