#' wellformedReadFilter
#' @description Filter alleles based on adjacent indels/tags
#'
#' @param maf An MAF data frame, generated by \code{\link{vcfToMAF}} function.
#' @return An MAF data frame after filtering for adjacent alleles.
#' 
#' @import stringr


wellformedReadFilter <- function(maf, vcfHeader){
  i <- 1
  maf$Start_Position <- as.numeric(maf$Start_Position)
  maf$End_Position <- as.numeric(maf$End_Position)
  while (i >= 1 & i <= nrow(maf)){
    # ValidAlignmentStartReadFilter
    if (maf[i, 'Start_Position'] == 0){
      stop(paste0('Variant ', as.character(i), 
                  ' has an invalid alignment start.'))
    }
    
    # ValidAlignmentEndReadFilter
    if(maf[i, 'End_Position'] < maf[i, 'Start_Position']){
      stop(paste0('Variant ', as.character(i), 
                  ' has an invalid alignment end.'))
    }
    
    # AlignmentAgreesWithHeaderReadFilter
    ## contig exists and start is within its range
    if (grep(paste0('contig.{1,}', maf[i, 'Chromosome'], ','), vcf_header[,1])){
      inLine <- grep(paste0('contig.{1,}', maf[i, 'Chromosome'], ','), 
                     vcf_header[, 1])
      contig <- as.numeric(str_extract(vcf_header$`Anno_Vcf@meta`[inLine], 
                                       '(?<=(length=))[:digit:]+'))
      if(maf[i, 'Start_Position'] > contig) {
        stop(paste0('Chromosme information of variant ', as.character(i), 
                    ' doesn\'t align with contig.'))
      }
    }else{
      stop(paste0('Chromosme information of variant ', as.character(i), 
                  ' doesn\'t align with contig.'))
    }
    
    # CigarContainsNoNOperator
    if(any(str_detect(maf[i, c('Reference_Allele', 'Tumor_Seq_Allele2')], 'N'))){
      stop(paste0('Variant ', as.character(i), ' contains skipped region(s).'))
    }
    
    i <- i + 1
  }
}